src/valkeymodule.h

```diff
@@ -794,6 +795,87 @@ typedef void (*ValkeyModuleInfoFunc)(ValkeyModuleInfoCtx *ctx, int for_crash_rep
 typedef void (*ValkeyModuleDefragFunc)(ValkeyModuleDefragCtx *ctx);
 typedef void (*ValkeyModuleUserChangedFunc)(uint64_t client_id, void *privdata);
 
+/* Current ABI version for scripting engine modules. */
+#define VALKEYMODULE_SCRIPTING_ENGINE_ABI_VERSION 1UL
+
+/* Type definitions for implementing scripting engines modules. */
+typedef void ValkeyModuleScriptingEngineCtx;
+typedef void ValkeyModuleScriptingEngineFunctionCtx;
+
+/* This struct represents a scripting engine function that results from the
+ * compilation of a script by the engine implementation.
+ *
+ * IMPORTANT: If we ever need to add/remove fields from this struct, we need
+ * to bump the version number defined in the
+ * `VALKEYMODULE_SCRIPTING_ENGINE_ABI_VERSION` constant.
+ */
+typedef struct ValkeyModuleScriptingEngineCompiledFunction {
+    char *name;       /* Function name */
+    size_t name_len;  /* The length of the function name string */
+    void *function;   /* Opaque object representing a function, usually it'
+                         the function compiled code. */
+    char *desc;       /* Function description */
+    size_t desc_len;  /* The length of the description string */
+    uint64_t f_flags; /* Function flags */
+} ValkeyModuleScriptingEngineCompiledFunction;
```

Comment:
@PingXie These `name` and `desc` are the strings that were sds before. We can't use sds in the module API, so we have `char *` and `ValkeyModuleString` to choose between.

@rjd15372 you mentioned that these can potentially contain NUL bytes, which is why you needed to add `name_len` and `desc_len`. This seems like a reason to use `robj` (AKA ValkeyModuleString) for these instead of `char *`. They're not in any hot path so the overhead of an robj isn't a problem in this case. Shall we close on this?

Reply:
I'm fine with using robj here. I'll update the PR soon.

Reply:
done

Reply:
@PingXie These are the names and descriptions of a *function*, not the engine itself. The lifetime is not that of the module, as functions and libraries can be deleted and redefined using the FUNCTION subcommands. The function names are strings that come from the Lua interpreter (in the Lua case). They can contain NUL characters. There's even a test for that. That's why we needed a `size_t length` along with the char pointers, but now it's changed to an `robj` object instead.

The following Lua code passed to FUNCTION CREATE results in a function with the name "my_hset" and the lib name "mylib":

```lua
#!lua name=mylib

local function my_hset(keys, args)
  local hash = keys[1]
  local time = server.call('TIME')[1]
  return server.call('HSET', hash, '_last_modified_', time, unpack(args))
end

server.register_function('my_hset', my_hset)
```

Reply:
I see. Would it be simpler to just duplicate these C strings as opposed to refcounting them via robj?

Reply:
@PingXie that was the initial implementation before changing it in commit 4a06209e . I don't mind to revert back to the original implementation where I use a char* + size_t to store these names.

Reply:
Yeah I am good with `robj`. No need to revert.